CPEX0046 Particle extension

araoCSI opened the pull request:
https://github.com/CGNS/CGNS/discussions/766

MicK7 commented on src/tests/test_particle.c:
```diff
@@ -0,0 +1,294 @@
+
+          for(int kk = 0; kk < size; ++kk)
+          {
+             if(coord_x[kk] != water_xc[kk] || coord_y[kk] != water_yc[kk] || coord_z[kk] != water_zc[kk])
```
Should use **compareValuesFloat** like in test_back_comp.c

MicK7 commented on src/tests/test_particle.c:
```diff
@@ -0,0 +1,294 @@
+
+          for(int kk = 0; kk < size; ++kk)
+          {
+             if(coord_x[kk] != water_xc[kk] || coord_y[kk] != water_yc[kk] || coord_z[kk] != water_zc[kk])
```
Should use **compareValuesFloat** like in test_back_comp.c

MicK7 commented on src/tests/test_particle.c:
```diff
@@ -0,0 +1,294 @@
+
+          for(int kk = 0; kk < size; ++kk)
+          {
+             if(coord_x[kk] != water_xc[kk] || coord_y[kk] != water_yc[kk] || coord_z[kk] != water_zc[kk])
```
Should use compareValuesFloat like in test_back_comp.c

MicK7 commented on src/tools/cgnscheck.c:
```diff
@@ -5971,6 +5990,569 @@
+    char name[33];
+    int n, nf;
+    int ndim;
+    cgsize_t datasize, size, dims[12];
```
please remove unused variable

MicK7 commented on src/tests/test_particle.c:
```diff
@@ -0,0 +1,294 @@
+    for(int iparticle = 1; iparticle <= nparticles; ++iparticle)
+    {
+       char name[32];
+       cgsize_t size;
```
to fix

MicK7 commented on src/tests/test_particle.c:
```diff
@@ -0,0 +1,294 @@
+          for(int ifield = 1; ifield <= nfields; ++ifield)
+          {
+             CGNS_ENUMT(DataType_t) type;
+             char name[32];
```
to fix

MicK7 commented:
Please rebase on latest develop branch

MicK7 commented:
Thanks for taking the time to improve the MR. There are still some glitches

AlexandreMinotCSI commented:
We will work on CGNS-modern instead

araoCSI commented:
@MicK7 Thanks for the initial review. I've updated the code to include doxygen comments wherever applicable. The sphinx version of the docs have also been pushed to it's repo. Please let me know if anything else is needed from my end.

brtnfld commented:
- [ ] Resync PR with develop
- [ ] Switch from Sphinx Docs of new functions to Doxygen.

MicK7 commented:
@brtnfld I tried to resync this PR with develop. It will help to have a cleaner diff for review.
@araoCSI Could you check that everything is still good for you with this PR ?

araoCSI commented:
@MicK7 Thanks for rebasing it. I've verified that everything works as expected

MicK7 commented:
@brtnfld The sphinx doc seems almost complete. I just haven't checked that it renders properly.

MicK7 commented:
Just add some time to check the doxygen and it's broken.

Missing in doxygen config :
`ALIASES += P_ParticleZone="Particle Zone index number, where 1 ≤ \e P ≤ \e n_particle_zones." `

`cg_particle_id` is not documented while being a public API.

Definition of param is not coherent between doc and API (ie file_number instead of fn is used) , that should be easy to fix

We are close to the end line.

araoCSI commented:
> Just add some time to check the doxygen and it's broken.
> 
> Missing in doxygen config : `ALIASES += P_ParticleZone="Particle Zone index number, where 1 ≤ \e P ≤ \e n_particle_zones." `
> 
> `cg_particle_id` is not documented while being a public API.
> 
> Definition of param is not coherent between doc and API (ie file_number instead of fn is used) , that should be easy to fix
> 
> We are close to the end line.

@MicK7 Thanks for spotting this. Is there some build flag or test that I can use to catch such inconsistencies? If not, what would be the best way to go about finding these?

MicK7 commented:
> > Just add some time to check the doxygen and it's broken.
> > Missing in doxygen config : `ALIASES += P_ParticleZone="Particle Zone index number, where 1 ≤ \e P ≤ \e n_particle_zones." `
> > `cg_particle_id` is not documented while being a public API.
> > Definition of param is not coherent between doc and API (ie file_number instead of fn is used) , that should be easy to fix
> > We are close to the end line.
> 
> @MicK7 Thanks for spotting this. Is there some build flag or test that I can use to catch such inconsistencies? If not, what would be the best way to go about finding these?

You need to run `doxygen doxygen/doxygen.config` in the src directory

MicK7 commented:
@brtnfld The doxygen issue should be solved now.
Better check twice...

MicK7 commented:
@brtnfld Any chance to get a review this week on your side ?

brtnfld commented:
Thanks for the reminder. I'll review it this week.

brtnfld commented on src/cg_ftoc.c:
```diff
@@ -2546,6 +2546,490 @@ CGNSDLL void cg_1to1_average_write_f(
+/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - *\
+ *   Read and write ParticleZone_t Nodes                                 *
+\* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
+
```
FYI, we are moving away from "C wappers" for Fortran and instead calling the C APIs directly from Fortran (cgns_f.F90). If it is easy for you to do that, it is preferred; otherwise, I can do it. I don't want it to hold up the PR.

I need to put a developer's note in cg_ftoc.c noting this policy,

brtnfld commented on src/cgns_f.F90:
```diff
@@ -550,6 +550,98 @@ END SUBROUTINE cgio_get_dimensions_f_1
 !DEC$ATTRIBUTES DLLEXPORT :: ModelTypeName
 !DEC$endif
 
+CHARACTER(LEN=MAX_LEN) :: ParticleGoverningEquationsTypeName(0:4)
```
It should start at 1, not 0.

brtnfld replied:
Nvm, I see the others start at 0, as well.

brtnfld commented on src/cgns_header.h:
```diff
@@ -510,6 +527,38 @@ typedef struct {            /* FlowEquationSet_t Node               */
+    CGNS_ENUMT(ParticleGoverningEquationsType_t) type;
+    int ndescr;             /* no of Descriptor_t nodes             */
+    cgns_descr *descr;      /* ptrs to in-memory copy of descr      */
+    int nuser_data;         /* number of user defined data nodes    */  /* V2.1 */
```
@MicK7 , Do you know if we need to indicate the version here? This is a new struct, and obviously, older versions of CGNS are not applicable. I'm assuming this was included to suggest what version the new field was added.

MicK7 replied:
I don't think we need the version here since this kind of node is already available but it would be nice to add a version for the top structure  to  tell that `ParticleGovervinvEquation_t`  was introduced in  `/* V4.5 */` 
I need to check but I think that something similar was done for the FamilyTree CPEX.

brtnfld commented on src/cgns_header.h:
```diff
@@ -918,6 +1024,7 @@ typedef struct {            /* CGNSBase_t Node          */
     CGNS_ENUMT(DataClass_t) data_class; /* Class of data                        */
     cgns_units *units;      /* Dimensional Units            */
     cgns_equations *equations; /* ptrs to in-mem. copy of FlowEqu.  */
+    cgns_pequations *pequations; /* ptr to in-mem. copy of ParticleEqu. */
```
@MicK7, For example, here, I think we might add a version of when it was introduced. This would also be useful for the Doxygen docs.

MicK7 replied:
I fully agree we need the version here.

brtnfld commented on src/cgns_internals.c:
```diff
@@ -6759,161 +6797,686 @@ cgns_link *cgi_read_link (double node_id)
     return CG_OK;
 }
 
```
I don't think this diff is getting rendered correctly.

brtnfld commented on src/cgnslib.c:
```diff
@@ -13528,226 +13644,1865 @@ int cg_1to1_average_write(int fn, int B, int Z, int J,
 }
 
```
Same here; this is not rendering correctly. I'll have to review it another way.

brtnfld reviewed (changes requested):
These are just some of the APIs that don't have testing associated with them. We can only release APIs that have testing, as we need to test them on various compilers and OSs.

cg_piter_write, cg_piter_read *, cg_piter_read* (* Are these exercised in the tools check?)
cg_particle_id,  cg_particle_coord_node_read, cg_particle_coord_node_write
cg_particle_bounding_box_read, cg_particle_bounding_box_write (verify the APIs do the correct thing if the datatype differs from the one in the file)
cg_particle_coord_info, cg_particle_coord_general_read
cg_particle_coord_id, cg_particle_coord_partial_write
cg_particle_coord_general_write

and some of the parallel APIs:

cgp_particle_coord_multi_read_data, cgp_particle_coord_multi_write_data

brtnfld commented on src/pcgnslib.c:
```diff
@@ -1968,161 +1968,172 @@ int cgp_field_general_read_data(int fn, int B, int Z, int S, int F,
         dataset, CG_PAR_READ);
 }
 
```
*sigh*, GitHub is messing up the diffs here.

brtnfld commented on src/cgns_internals.c:
It looks good after reviewing it with Meld.

brtnfld commented on src/tests/test_particle.c:
As noted in the general comments, we need to test all new public APIs.

brtnfld commented on src/tools/cgnscheck.c:
```diff
@@ -5972,10 +5991,571 @@ static void check_base_iter (void)
+
+   strcpy (p->name, name);
+   p->nparticles = size[0];
+   // 0 sized ParticleZones are okay as they indicate that there are no parcels of this type in the Base at the current time
```
sp. parcels

araoCSI replied:
Fixed

brtnfld commented on src/tools/cgnscheck.c:
```diff
@@ -5971,6 +5990,569 @@
+        if (verbose)
+            printf("      Coordinate Range=%g -> %g (%g)\n",
+                cmin, cmax, cmax-cmin);
+        if (cmin == cmax)
```
Should address.

brtnfld commented on src/cgnslib.c:
```diff
@@ -13528,226 +13644,1865 @@ int cg_1to1_average_write(int fn, int B, int Z, int J,
- *
- * \brief Access a node via pathname
- *
+   /* Verify the particle count - We allow 0 sized ParticleZone_t nodes but they cannot have and solution/coordinate arrays */
```
Verify the particle count - we allow 0-sized ParticleZone_t nodes, but they cannot have solution/coordinate arrays

araoCSI replied:
Good catch, fixed

brtnfld commented:
Could you please clarify if you intended to support particles in the ADF* format?

brtnfld commented on src/cgnslib.c:
```diff
@@ -13528,226 +13644,1865 @@ int cg_1to1_average_write(int fn, int B, int Z, int J,
+   strcpy(solname, sol->name);
+   return CG_OK;
+}
+
```
Missing doxygen.

araoCSI replied:
Done

brtnfld commented on src/cgnslib.c:
I reviewed using Meld.

brtnfld commented on src/pcgnslib.c:
```diff
@@ -1968,161 +1968,172 @@ int cgp_field_general_read_data(int fn, int B, int Z, int S, int F,
+ * \param[in] coordname \PGRID_coordname
+ * \param[out] C \PGRID_Coordinate
  * \return \ier
+ * \details To write the data in parallel, first call /e cgp_coord_write to create an empty data node. This call is identical
```
use other slash, it's "\e" and "\p"

araoCSI replied:
Done

brtnfld commented:
Complete source code review.

MicK7 commented:
@araoCSI can you take care of Scot review please. At least for the testing part.

araoCSI commented:
@brtnfld Thanks for the detailed review. I've added tests for all the new API functions to test_particle.c (serial) and pcgns_test.c (parallel). Please let me know if you'd like me to split up the tests (i.e. add new files to have more categorization). I've also updated the version numbers for the new structs in cgns_header.h.

As far as the Fortran changes go, I added the routines as part of the original commit to cgns_f.F90. I'm not sure if this is what you meant when you said "calling the C APIs directly from Fortran (cgns_f.F90)". Quite a lot of them were left commented since their equivalent Zone_t routines were also commented. If more changes need to be made, I'd be okay with doing it but I'll definetely need some guidance from you haha.

Please let me know if any other changes need to be made.

MicK7 commented on src/tests/test_particle.c:
```diff
@@ -0,0 +1,754 @@
+             char fname[32];
+             cg_particle_field_info(fnum, bnum, iparticle, isol, ifield, &type, fname);
+
+             cgsize_t rmin = 1, rmax = size;
```
```suggestion
             rmin = 1; rmax = size;
```

araoCSI replied:
Fixed

brtnfld commented on src/tests/test_particle.c:
```diff
@@ -0,0 +1,754 @@
+    /* It is okay to define ParticleZone_t nodes with 0 particles
+     * this implies that for the current solution time, no particles
+     * of said type (IsoOctane and Oil) are in the Base */
+    cg_particle_write(fnum, bnum, "IsoOctane", 0, &pnum); // 0 IsoOctane particles
```
Can you add pass-fail checks for the new particle APIs?

araoCSI replied:
@brtnfld By pass-fail checks, did you mean that I should use write_test_status (similar to pcgns_ctest.c)?

brtnfld replied:
I meant to check the API's return status to ensure it did not fail.

araoCSI replied:
Those should get passed to the error call back function (print_error) and printed to the screen right? We have 1 test in that file where we expect an error (trying to create a particle bbox with ints, but this shouldn't be fatal). Fatal errors in any API function i.e. code -1 will cause the test to fail (exit(1)).  I can remove this call back and put all function inside an if statement if that is preferred.

brtnfld replied:
I'm sorry; I missed the cg_configure. Yes, a callback is OK. However, that will not work with the test status schema unless you print the test header and then print FAIL in the callback. I don't know what that will gain except that it would indicate at what point in the test it fails. It is fine as is.

brtnfld commented on src/tools/cgnscheck.c:
```diff
@@ -5972,10 +5991,571 @@ static void check_base_iter (void)
+        if (verbose)
+            printf("      Coordinate Range=%g -> %g (%g)\n",
+                cmin, cmax, cmax-cmin);
+        if (cmin - cmax == 0)
```
Or use the compareValues* function.

araoCSI replied:
Done

brtnfld commented:
> @brtnfld Thanks for the detailed review. I've added tests for all the new API functions to test_particle.c (serial) and pcgns_test.c (parallel). Please let me know if you'd like me to split up the tests (i.e. add new files to have more categorization). I've also updated the version numbers for the new structs in cgns_header.h.

It would be ideal to have progression test outputs with pass/fail to identify where the test failed more easily. However, one file is OK.

> 
> As far as the Fortran changes go, I added the routines as part of the original commit to cgns_f.F90. I'm not sure if this is what you meant when you said "calling the C APIs directly from Fortran (cgns_f.F90)". Quite a lot of them were left commented since their equivalent Zone_t routines were also commented. If more changes need to be made, I'd be okay with doing it but I'll definetely need some guidance from you haha.

It's fine as is, but you can look at  cgp_poly_section_write_f as an example. There is no C equivalent "wrapper" that is called from Fortran.  Some of the new APIs are easy to do, and some are not so much. If you don't get around to making Fortran tests for the latest APIs, I can do the conversion when I work on the test.

Thanks for all the hard work you've put in.

> 
> Please let me know if any other changes need to be made.

brtnfld commented:
To clarify, you can use the write_test_* functions in utils.h to indicate the category being tested and whether it passes or fails as the test progresses.

araoCSI commented:
> Could you please clarify if you intended to support particles in the ADF* format?

@brtnfld Sorry, I didn't see this earlier. If we plan on supporting ADF* format for the forseeable future, I don't see any reason as to why we shouldn't support particles in this format. If required, I can set the format to ADF for one of the particle tests.

I also ran cgnscheck against the output files produced by the tests and fixed a few more bugs. So now, cgnscheck is also clean (i.e. has expected warnings for Family_t nodes but no errors). I've not modified test_particle.c to use the schema but I've added print statements which show the progress of the tests and I've verified that fatal errors cause the test to fail as expected.

>It's fine as is, but you can look at cgp_poly_section_write_f as an example. There is no C equivalent "wrapper" that is called from Fortran. Some of the new APIs are easy to do, and some are not so much. If you don't get around to making Fortran tests for the latest APIs, I can do the conversion when I work on the test.

I'm unable to compile the library when I turn on CGNS_ENABLE_FORTRAN and CGNS_ENABLE_PARALLEL. The config/generation steps work fine but I end up with this error (corresponds to `c_offsets = C_LOC(offsets)` in cgp_poly_elements_write_data_f)

`Error: Assumed-shape array 'elements' at (1) cannot be an argument to the procedure 'c_loc' because it is not C interoperable`

I'm not quite sure if this is an issue with my compiler/configuration or an issue with the library. I've not added any new fortran code as I cannot build/test it. Sorry about that!

brtnfld commented:
> @brtnfld Sorry, I didn't see this earlier. If we plan on supporting ADF* format for the forseeable future, I don't see any reason as to why we shouldn't support particles in this format. If required, I can set the format to ADF for one of the particle tests.

We have not been back-porting newer CGNS features, such as 64-bit or a variety of element types. What happens if you switch to ADF2?

brtnfld commented:
> I'm unable to compile the library when I turn on CGNS_ENABLE_FORTRAN and CGNS_ENABLE_PARALLEL. The config/generation steps work fine but I end up with this error (corresponds to `c_offsets = C_LOC(offsets)` in cgp_poly_elements_write_data_f)
> 
> `Error: Assumed-shape array 'elements' at (1) cannot be an argument to the procedure 'c_loc' because it is not C interoperable`
> 
> I'm not quite sure if this is an issue with my compiler/configuration or an issue with the library. I've not added any new fortran code as I cannot build/test it. Sorry about that!

This is the first time I've seen that issue. What compiler/version and OS are you using?

brtnfld commented:
That is an ancient compiler, so I'm not surprised there are issues; plus, Centos 7 is EOL, so we don't bother supporting it.

brtnfld commented:
@araoCSI, are you done with the changes?
@MicK7, do you see any remaining tasks that need to be completed before merging?

FYI, once this is merged, another, more extensive compiler/platform testing will occur, so some additional issues may arise.

araoCSI commented:
@brtnfld @MicK7  Yes. To the best of my knowledge, all new APIs and tests work as expected. Thanks for taking the time to review these changes.

MicK7 commented:
@brtnfld This PR looks in good shape. I don't see any other things to do (outside of taking care of cg_ftoc.c but it can be done in a separate PR to cope with all fortran API at the same time) .
